Trac 50617

claytoncollie opened the pull request:
Code and tests for adding an additional query parameter to the REST API request. 

Added two new arguments, `modified_before` and `modified_after` which request data based on the `post_modified` field from the WP_Post object.

Trac ticket: https://core.trac.wordpress.org/ticket/50617

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

claytoncollie committed:
Add modified_before and modified_after to get_collection_params method

claytoncollie committed:
Add array keys and args into get_items method

claytoncollie committed:
Add proper columns for post_modified meta value

claytoncollie committed:
Add guard so we dont run 'after' and 'modified_after' or 'before' and 'modified_before' at the same time

claytoncollie committed:
Formatting

claytoncollie committed:
Add comments about guarding against using 'before' or 'after' in the same request as these new collection params

claytoncollie committed:
Add collection params for modified_before and modified_after to Comments controller

claytoncollie committed:
Add array keys and args into get_items method for Comments Controller

claytoncollie committed:
Guard against running 'before' and 'after' along side 'modified_before' and modified_after' in the same request

claytoncollie committed:
Add test for invalid and valid modified date for Posts Controller

claytoncollie committed:
Add test for invalid and valid modified date for Pages Controller

claytoncollie committed:
Format test methods to WordPress Coding Standards

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -251,11 +251,25 @@ public function get_items( $request ) {
 
+		// Set modified_before into date query. Date query must be specified as an array of an array.
+		// Guard against using 'before' and 'modified_before' in the same request.
+		if ( isset( $registered['modified_before'], $request['modified_before'] ) && ! isset( $request['after'] ) ) {
```
Can we issue a `400` error in this case instead? I think I'd found it unintuitive as a developer if one of them were silently ignored.

The same in the other checks as well.

claytoncollie replied:
Added to Posts and Comments controller.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/rest-attachments-controller.php:
```diff
@@ -566,6 +566,51 @@ public function test_get_items_valid_date() {
 		$this->assertEquals( $id2, $data[0]['id'] );
 	}
 
+	public function test_get_items_invalid_modified_date() {
```
```suggestion
	/*
	 * @ticket 50617
	 */
	public function test_get_items_invalid_modified_date() {
```

`@ticket` annotations should be added for any new tests introduced.

claytoncollie replied:
Added the doc block to all new tests.

TimothyBJacobs reviewed (changes requested):
Thanks for the patch @claytoncollie!

A couple of minor bits of feedback.

claytoncollie committed:
Add docblock for new tests with ticket number

claytoncollie committed:
Add guard clause to Comments controller. Do not allow modified_before/after to be queried alongside after/before.

claytoncollie committed:
Add guard clause to Posts controller. Do not allow modified_before/after to be queried alongside after/before.

claytoncollie commented on src/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -251,11 +251,41 @@ public function get_items( $request ) {
 
+		// Guard against using 'before' and 'modified_before' in the same request.
+		if ( isset( $registered['modified_before'], $request['modified_before'] ) && isset( $request['before'] ) ) {
+			return new WP_Error(
```
How does this look for a guard clause?

claytoncollie reviewed (commented):
Added the docblock to the test methods with the ticket number.

Added guard clauses to return WP_Error when modified_before/after are used in conjunction with after/before.

TimothyBJacobs commented:
Sorry for the back and forth on this @claytoncollie. `WP_Date_Query` does support an array of clauses, so why can't we support searching by both columns at once actually?

claytoncollie commented:
@TimothyBJacobs Not a problem at all. I want to get the right so I appreciate you looking into it. I looked at `WP_Date_Query` and the related tests where a person is querying by post_date and post_modified date at the same time. The first date array is built with the post_date data then a second array is created to house the post_modified data. The reason I put the conditional in my PR is that we are setting the first array `$args['date_query'][0]` in both instances. 

If you like, I can write some more code to support where they work individually and where they work together. This is the first pass but it feels weird. It also doesn't support when somebody is querying by `after` and `modified_before` at the same time. Only `after and `modified_after` and then `before` and `modified_before`. The combinations start to get out of hand.

```php
// Set modified_before into date query. Date query must be specified as an array of an array.
if ( isset( $registered['modified_before'], $request['modified_before'] ) ) {
	// Check if we are using modified_before and before in the same query.
	if ( isset( $registered['before'], $request['before'] ) ) {
		$args['date_query'][1]['before'] = $request['modified_before'];
		$args['date_query'][1]['column'] = 'post_modified';
	} else {
		$args['date_query'][0]['before'] = $request['modified_before'];
		$args['date_query'][0]['column'] = 'post_modified';
	}
}
```

and

```php
// Set modified_after into date query. Date query must be specified as an array of an array.
if ( isset( $registered['modified_after'], $request['modified_after'] ) ) {
	// Check if we are using modified_after and after in the same query.
	if ( isset( $registered['after'], $request['after'] ) ) {
		$args['date_query'][1]['after'] = $request['modified_after'];
		$args['date_query'][1]['column'] = 'post_modified';
	} else {
		$args['date_query'][0]['after'] = $request['modified_after'];
		$args['date_query'][0]['column'] = 'post_modified';
	}
}
```

TimothyBJacobs commented:
Sorry for the delay in responding to your comment @claytoncollie.

Instead of trying to implement it in one clause, can we make each date check its own clause? Something like this?

```php
$date_query = array();

if ( isset( $registered['before'], $request['before'] ) ) {
	$date_query[] = array(
		'before' => $request['before'],
		'column' => 'post_date',
	);
}

if ( isset( $registered['modified_before'], $request['modified_before'] ) ) {
	$date_query[] = array(
		'before' => $request['modified_before'],
		'column' => 'post_modified',
	);
}

if ( isset( $registered['after'], $request['after'] ) ) {
	$date_query[] = array(
		'after'  => $request['after'],
		'column' => 'post_date',
	);
}

if ( isset( $registered['modified_after'], $request['modified_after'] ) ) {
	$date_query[] = array(
		'after'  => $request['modified_after'],
		'column' => 'post_modified',
	);
}
```

claytoncollie committed:
Remove guard clauses to allow date_query by post_date and post_modified at the same time on Comments

claytoncollie committed:
Remove guard clauses to allow date_query by post_date and post_modified at the same time on Posts

claytoncollie committed:
Revert change to Comments controller because WP_Comment does not support modified date.

claytoncollie committed:
Add new parameters to Post, Page and Attachment test fixtures for when testing query params

claytoncollie commented:
Thanks for getting back to me @TimothyBJacobs 

Your code is much more simple and builds the array properly. I updated the Post Controller with this code and it works perfectly.

I also removed the code I had for the Comments Controller as the WP_Comment class does not have a modified date as a class parameter.

The last thing I did with this PR is to add the `modified_after` and `modified_before` array keys to the test fixtures for testing the query params on Post, Page, and Attachments.

claytoncollie committed:
Update test cases to start with post_date in the past, then update the post_modified date to be a few days into the future.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/rest-attachments-controller.php:
```diff
@@ -566,6 +568,60 @@ public function test_get_items_valid_date() {
+				'post_excerpt'   => 'A sample caption',
+			)
+		);
+		$this->update_post_modified( $id1, '2016-01-15T00:00:00Z' );
```
```suggestion
		$this->update_post_modified( $id1, '2016-01-15 00:00:00' );
```

This function accepts dates as `YYYY-MM-DD HH:MM:SS`.

claytoncollie replied:
I changed all of my dates to follow the correct format.

TimothyBJacobs replied:
Looks good!

TimothyBJacobs commented on tests/phpunit/tests/rest-api/rest-attachments-controller.php:
```diff
@@ -566,6 +568,60 @@ public function test_get_items_valid_date() {
+			$this->test_file,
+			0,
+			array(
+				'post_date'      => '2016-01-01T00:00:00Z',
```
This should also be doing `YYYY-MM-DD HH:MM:SS`.

claytoncollie replied:
I changed all of my dates to follow the correct format.

claytoncollie committed:
Update date format for Posts

claytoncollie committed:
Update date format for Pages

claytoncollie committed:
Update date format for Attachments

TimothyBJacobs commented:
Merged in https://core.trac.wordpress.org/changeset/50024.

ak1051210 commented:
I acknowledge your proposal to enhance the WP_REST_Request attributes by adding modified_before and modified_after to enable querying posts by their modified date. This addition would complement the existing before and after attributes, offering greater flexibility in querying posts through the REST API.
{{{
#!html
<a href="https://tassel.pk"></a>
<a href="https://aklasbelakarachi.com"></a>
<a href="https://dailylasbelagamekarachi.com"></a>
<a href="https://cheapcamshows.com"></a><a href="https://www.nogentech.org"></a><a href="https://www.edulize.com"></a><a href="https://www.webmastershall.com"></a><a href="https://www.techuy.com"></a><a href="https://justmyfitness.com"></a><a href="https://stuffablog.com"></a><a href="https://www.bravotech.org"></a><a href="https://tricksmode.com"></a>

}}}
